# Drive failure from 30-day SMART windows — Backblaze, 2022 to today, CNN vs XGBoost

The same pipeline as `Dataset_backblaze_Analysis_2022.ipynb` — archives from **2022
onward** — with a second model added: **XGBoost** on a flattened summary of each 30-day
window, scored against the CNN on the same test drives and the same windows.

| § | question | what gets printed |
|---|---|---|
| **1** | is any column a proxy for `failure`? | the leakage audit and the surviving feature set $X$ |
| **2** | can the model see the label at all? | **four executed checks**, not a claim |
| **3** | how many 30-day windows cover a drive's whole life? | per-drive window and **batch counts** |
| **4** | does this drive fail in *this* 30 days? (CNN) | a verdict per window, plus per-drive histograms |
| **5** | does gradient boosting on window summaries beat the CNN? | XGBoost training, test metrics, **CNN vs XGBoost** table and curves |
| **6** | what does a tree grown to 100% accuracy look like? | the depth ladder, and what it costs |
| **Summary** | which model is best? | all three models in one table, and the winner by drive precision @100 and PR-AUC |

**Scope.** Archives `Q1_2022` through the newest quarter. A drive that was already in
service in 2021 enters with only its 2022-onward rows, which is the intended reading of
"from 2022 till today".

**Still the fleet, not a cohort.** No vendor filter and no matched sampling: failures are
a fraction of a percent of drive-days, as they are in a real datacentre.

In [1]:
import gc, json, os, sys, time
import numpy as np
import pandas as pd
import torch
import xgboost as xgb
import matplotlib.pyplot as plt

sys.path.insert(0, "scripts")
import leakage_audit as la
import backblaze_window_pipeline as bw
import archive_window_pipeline as ap

plt.rcParams.update({"figure.dpi": 110, "axes.grid": True, "grid.alpha": 0.25,
                     "font.size": 9})
pd.set_option("display.width", 220, "display.max_columns", 40)

SHARD_DIR  = os.environ.get("BACKBLAZE_SHARDS",
                            os.path.join("data", "backblaze_archive", "shards"))
STORE_DIR  = os.environ.get("BACKBLAZE_STORE_2022",
                            os.path.join("data", "backblaze_archive", "store_2022"))
SINCE_YEAR = int(os.environ.get("SINCE_YEAR", "2022"))
FIG_DIR    = "results/figures"
SEED       = 42
os.makedirs(FIG_DIR, exist_ok=True)
bw.set_seed(SEED)

shards = ap.shard_paths(SHARD_DIR, since_year=SINCE_YEAR)
print(f"torch {torch.__version__} | xgboost {xgb.__version__} | device "
      f"{'cuda' if torch.cuda.is_available() else 'cpu'}")
print(f"{len(shards)} shards from {SINCE_YEAR} on: "
      f"{os.path.basename(shards[0])[:-8]} -> {os.path.basename(shards[-1])[:-8]}")
print("  " + ", ".join(os.path.basename(p)[:-8] for p in shards))

torch 2.14.0.dev20260708+cu132 | xgboost 3.4.1 | device cuda
18 shards from 2022 on: Q1_2022 -> Q2_2026
  Q1_2022, Q2_2022, Q3_2022, Q4_2022, Q1_2023, Q2_2023, Q3_2023, Q4_2023, Q1_2024, Q2_2024, Q3_2024, Q4_2024, Q1_2025, Q2_2025, Q3_2025, Q4_2025, Q1_2026, Q2_2026


---
## 1 — Target-leakage audit

A leaked column is not *evidence about* the outcome but a *restatement of* it. In drive
telemetry it is almost never a SMART attribute; it hides in three places, and the audit
probes all three: the column values, the **missingness** pattern, and **record
geometry** — anything derived from where a row sits relative to the end of a drive's
record, because Backblaze writes `failure = 1` on a drive's last reporting day.

The audit samples by **drive**, never by drive-day: every geometry probe is derived from
a drive's own first and last day, and a drive whose rows have been thinned has neither.

In [2]:
t0 = time.time()
cfg0 = ap.ArchiveConfig(shard_dir=SHARD_DIR, store_dir=STORE_DIR,
                        since_year=SINCE_YEAR, random_state=SEED)
census = ap.build_drive_census(cfg0)

epoch = pd.Timestamp("1970-01-01")
print(f"\n{len(census):,} drives | {int(census.n_obs.sum()):,} observed drive-days | "
      f"{int(census.has_failure.sum()):,} drives with a failure event")
print(f"{epoch + pd.Timedelta(days=int(census.start_day.min())):%Y-%m-%d} -> "
      f"{epoch + pd.Timedelta(days=int(census.end_day.max())):%Y-%m-%d}"
      f"   ({time.time()-t0:.0f}s)")

  census cached: 416,428 drives

416,428 drives | 457,380,330 observed drive-days | 19,024 drives with a failure event
2022-01-01 -> 2026-06-30   (0s)


In [3]:
audit_df = ap.sample_audit_frame(cfg0, census, n_drives=15_000)
report = la.audit_frame(audit_df)
report.print_report()

  [ 1/18] Q1_2022.parquet    669,411 rows kept


  [ 2/18] Q2_2022.parquet    691,592 rows kept


  [ 3/18] Q3_2022.parquet    735,216 rows kept


  [ 4/18] Q4_2022.parquet    768,652 rows kept


  [ 5/18] Q1_2023.parquet    767,257 rows kept


  [ 6/18] Q2_2023.parquet    780,868 rows kept


  [ 7/18] Q3_2023.parquet    843,212 rows kept


  [ 8/18] Q4_2023.parquet    885,087 rows kept


  [ 9/18] Q1_2024.parquet    902,083 rows kept


  [10/18] Q2_2024.parquet    933,665 rows kept


  [11/18] Q3_2024.parquet    962,131 rows kept


  [12/18] Q4_2024.parquet    985,769 rows kept


  [13/18] Q1_2025.parquet  1,002,847 rows kept


  [14/18] Q2_2025.parquet  1,037,907 rows kept


  [15/18] Q3_2025.parquet  1,074,738 rows kept


  [16/18] Q4_2025.parquet  1,116,665 rows kept


  [17/18] Q1_2026.parquet  1,104,216 rows kept


  [18/18] Q2_2026.parquet  1,156,039 rows kept


  audit frame: 16,417,355 drive-days over 15,000 drives, 656 failure events (base rate 0.00400%)
[audit] scoring 23 telemetry columns, 6 identifier/topology columns, and 6 derived record-geometry probes over 16,417,355 drive-days ...


STEP 1 -- TARGET LEAKAGE AND PROXY AUDIT
  drive-days ........... 16,417,355
  distinct drives ...... 15,000
  failure events ....... 656 (base rate 0.0040%)

  thresholds: severe if rank-AUC >= 0.95, or 1-threshold balanced acc >= 0.90,
              or |r| >= 0.50, or missingness gap >= 0.25, or level lift >= 20x

--------------------------------------------------------------------------------------------
A. TELEMETRY COLUMNS vs failure -- top 12 by rank AUC
--------------------------------------------------------------------------------------------
       column coverage  n_unique   corr auc_directional balanced_acc verdict
smart_187_raw   0.3930      1120 0.0174          0.8887       0.8785 suspect
smart_183_raw   0.0325       169 0.0006          0.7950       0.7586   clean
smart_197_raw   0.9685      1396 0.0651          0.7862       0.7842   clean
  smart_5_raw   0.9910      6815 0.0212          0.7859       0.7807   clean
smart_198_raw   0.9902       890 0.0640          0.7004  

In [4]:
features = la.select_feature_columns(audit_df, report)

print(f"{len(features)} attributes survive coverage >= {report.cfg.min_coverage:.0%} "
      f"and the blocklist:\n")
tbl = report.columns.set_index("column")
for c in features:
    r = tbl.loc[c]
    print(f"  {c:<18} coverage {100*r.coverage:6.2f}%  "
          f"{int(r.n_unique):>9,} distinct  rank AUC {r.auc_directional:.4f}")

report.assert_clean(features)          # raises if a blocked name reached X
la.plot_audit(report, save_path=f"{FIG_DIR}/2022x_step1_leakage_audit.png")
del audit_df; gc.collect()

15 attributes survive coverage >= 50% and the blocklist:

  smart_197_raw      coverage  96.85%      1,396 distinct  rank AUC 0.7862
  smart_5_raw        coverage  99.10%      6,815 distinct  rank AUC 0.7859
  smart_198_raw      coverage  99.02%        890 distinct  rank AUC 0.7004
  smart_4_raw        coverage  98.77%        423 distinct  rank AUC 0.6590
  smart_12_raw       coverage  99.82%        346 distinct  rank AUC 0.6403
  smart_9_raw        coverage  99.82%     82,565 distinct  rank AUC 0.6211
  smart_240_raw      coverage  65.79%     87,644 distinct  rank AUC 0.6022
  smart_193_raw      coverage  98.67%    136,319 distinct  rank AUC 0.5959
  smart_1_raw        coverage  99.74%  5,427,754 distinct  rank AUC 0.5781
  smart_7_raw        coverage  98.77%  5,873,176 distinct  rank AUC 0.5710
  smart_3_raw        coverage  98.77%      3,417 distinct  rank AUC 0.5443
  smart_192_raw      coverage  99.49%     16,010 distinct  rank AUC 0.5293
  smart_194_raw      coverage  99.82%     

  saved results/figures/2022x_step1_leakage_audit.png


C:\Users\tempuser2\Desktop\XGBOOST_2.0\XGBOOST\scripts\leakage_audit.py:618: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


28

---
## 2 — Building the store, then **proving** the label cannot be read

The dense matrix is built once and memory-mapped. `failure` is written to a *separate
file* from the feature matrix — it is not a column of `X` that gets dropped later, it
never enters `X` at all.

That is a claim, so the next cell tests it four ways:

1. **Names.** No channel is called `failure` or derived from it.
2. **Storage.** `values.f32` and `failure.u8` are different files and share no memory, so
   no write to one can surface in the other.
3. **Independence.** The label array is replaced with random noise and the same windows
   are gathered again. If a single byte of `X` depended on the label, `X` would move. It
   must come back bit-identical.
4. **Separability.** Every channel is scored against the window label on its own. A
   column that *was* the label would sit at AUC ≈ 1.0. The check fails the notebook if
   any channel exceeds 0.95.

Checks 1, 2 and 4 raise on failure; check 3 is the one that actually settles it, because
it is an experiment on the running code rather than an inspection of it.

In [5]:
cfg = ap.ArchiveConfig(
    shard_dir         = SHARD_DIR,
    store_dir         = STORE_DIR,
    since_year        = SINCE_YEAR,
    window_days       = 30,     # a sample is 30 consecutive days
    horizon_days      = 30,     # positive if the drive fails within 30 days of window end
    stride_days       = 1,      # every possible start is enumerated for eval
    min_days          = 30,     # shorter drives are dropped, not padded
    delta_lags        = (1, 7),
    add_observed_mask = True,
    feature_columns   = tuple(features),
    log1p_columns     = tuple(c for c in features
                              if c not in ("smart_194_raw", "smart_190_raw")),
    n_buckets         = 48,
    samples_per_drive = 4,
    train_positive_ratio = 0.25,
    batch_size        = 4096,
    random_state      = SEED,
)

store = ap.build_archive_store(cfg)

STEP A -- DRIVE CENSUS
  census cached: 416,428 drives

STEP B -- DENSE LAYOUT


  413,045 drives kept, 3,383 dropped (shorter than 30 days)
  459,048,144 dense drive-days (457,317,880 observed, 1,730,264 forward-filled)
  matrix 459,048,144 x 46 float32 = 84.5 GB memory-mapped

STEP C -- EXTERNAL SORT INTO 48 BUCKETS
  buckets cached

STEP D -- DENSIFY
  dense matrix cached

ArchiveSeriesStore: 413,045 drives, 459,048,144 dense drive-days, 46 channels (84 GB memory-mapped), 18,564 drives with a failure event


In [6]:
splits = ap.grouped_drive_split(store, cfg)
checks = ap.verify_no_label_leakage(store, splits["train"], n_probe=20_000, seed=SEED)

LABEL-LEAKAGE VERIFICATION
  1. names ........... 46 channels, none named after `failure`
  2. storage ......... values -> values.f32, labels -> failure.u8 (separate files, no shared memory)
  3. independence .... labels overwritten with noise; X re-gathered bit-identical over 4,096 windows  PASS
  4. separability .... 20,000 probe windows, 0.395% positive
     strongest single channel: smart_197_raw at AUC 0.6911 (a proxy would sit at ~1.00)

  the model input is (batch, 30, 46) and contains SMART levels, their deltas and the observed mask -- nothing else


In [7]:
# The strongest channels, ranked against the window label. Useful signal sits well short
# of the 0.95 line; a proxy would be pinned against 1.00.
top = checks["channel_scores"].head(12).reset_index(drop=True)
top.index += 1
print("strongest single channels vs the 30-day window label "
      f"({checks['n_probe_windows']:,} probe windows):\n")
print(top.to_string())
print(f"\nceiling for a non-leaked channel: 0.95     observed maximum: "
      f"{checks['max_channel_auc']:.4f}")
print(f"model input shape: (batch, {cfg.window_days}, {checks['n_channels']}) "
      f"-- SMART levels, 1- and 7-day deltas, observed mask. No label channel.")

strongest single channels vs the 30-day window label (20,000 probe windows):

           channel  auc_vs_window_label
1    smart_197_raw             0.691124
2      smart_5_raw             0.682798
3    smart_198_raw             0.648322
4      smart_1_raw             0.627615
5      smart_7_raw             0.622210
6   smart_5_raw_d7             0.618829
7    smart_193_raw             0.602454
8      smart_3_raw             0.599101
9   smart_7_raw_d1             0.594618
10  smart_9_raw_d1             0.592584
11   smart_240_raw             0.590284
12  smart_7_raw_d7             0.590062

ceiling for a non-leaked channel: 0.95     observed maximum: 0.6911
model input shape: (batch, 30, 46) -- SMART levels, 1- and 7-day deltas, observed mask. No label channel.


---
## 3 — Covering every drive's whole timeline with 30-day windows

Two different questions, two different window sets, both reported below.

**Tiling** lays non-overlapping 30-day blocks end to end along each drive's record:
$\lceil \text{record days} / 30 \rceil$ windows, so **every drive-day lands in exactly
one window** — nothing counted twice, nothing skipped. This is "as many batches as it
takes to fill the whole timeline", and it is what the decision tree in §5 is fitted on.
Where a record is not a multiple of 30, the last block is pulled back to end on the
drive's final day, so it overlaps its predecessor rather than running short.

**Stride-1 enumeration** is the other extreme: every possible start offset, used for the
sequence model's validation and test passes.

Both counts are printed, per split and per drive.

In [8]:
tiles = {name: ap.TilingWindows(store, splits[name], batch_size=cfg.batch_size,
                                split_name=name)
         for name in ("train", "val", "test")}

W = cfg.window_days
rows = []
for name in ("train", "val", "test"):
    idx, t = splits[name], tiles[name]
    rows.append({
        "split": name,
        "drives": len(idx),
        "drives with a failure": int(store.drive_has_failure[idx].sum()),
        "record days (dense)": int(store.lengths[idx].sum()),
        f"tiling {W}-day windows": t.n_windows,
        "tiling batches": len(t),
        "every possible window": int((store.lengths[idx] - W + 1).sum()),
    })
coverage = pd.DataFrame(rows).set_index("split")
print(coverage.to_string(formatters={c: "{:,}".format for c in coverage.columns
                                     if coverage[c].dtype.kind in "iu"}))

total_tiles = sum(t.n_windows for t in tiles.values())
total_batches = sum(len(t) for t in tiles.values())
print(f"\nTIMELINE COVERAGE")
print(f"  {store.n_drives:,} drives, {store.n_rows:,} dense drive-days")
print(f"  {total_tiles:,} non-overlapping {W}-day windows cover the entire timeline")
print(f"  at batch size {cfg.batch_size:,} that is {total_batches:,} batches in total")
for name in ("train", "val", "test"):
    print("  " + tiles[name].label_report())

       drives drives with a failure record days (dense) tiling 30-day windows tiling batches every possible window
split                                                                                                             
train 280,870                12,623         312,175,229            10,519,256          2,569           304,029,999
val    49,566                 2,228          55,215,887             1,860,545            455            53,778,473
test   82,609                 3,713          91,657,028             3,088,649            755            89,261,367

TIMELINE COVERAGE
  413,045 drives, 459,048,144 dense drive-days
  15,468,450 non-overlapping 30-day windows cover the entire timeline
  at batch size 4,096 that is 3,779 batches in total
  train | tiling W=30  |   10,519,256 windows | 280,870 drives | positive rate  0.2401% (25,254 positive windows) | 2,569 batches
  val   | tiling W=30  |    1,860,545 windows |  49,566 drives | positive rate  0.2391% (4,449 positive wi

In [9]:
# Per-drive accounting: how many windows each drive needs to have its whole record
# covered, and how much the final window had to overlap to stay a full 30 days.
cov = tiles["test"].coverage_report()
print(f"per-drive window counts, first 20 test drives "
      f"(batch size {cfg.batch_size:,}):\n")
print(cov.head(20).to_string(index=False))

n = cov["windows_to_cover_timeline"]
print(f"\nacross all {len(cov):,} test drives:")
print(f"  windows per drive .... min {n.min()}, median {int(n.median())}, "
      f"mean {n.mean():.1f}, max {n.max()}")
print(f"  total windows ........ {int(n.sum()):,}")
print(f"  total batches ........ {len(tiles['test']):,} "
      f"of up to {cfg.batch_size:,} windows each")
print(f"  days covered ......... {int(cov.days_covered.sum()):,} "
      f"(record holds {int(cov.record_days.sum()):,}; the difference is the "
      f"overlap pulled back onto the last window of each drive)")

per-drive window counts, first 20 test drives (batch size 4,096):

 drive_index    serial_number  record_days  windows_to_cover_timeline  days_covered  overlap_on_last_window  drive_has_failure
           3 02308a3415be0010         1642                         55          1650                       8              False
           7 0559e3d009d20010          993                         34          1020                      27              False
           8 0564f6f3fab90010         1642                         55          1650                       8              False
           9 059bba587f390010          993                         34          1020                      27              False
          15 078ea1af2c3f0010         1642                         55          1650                       8              False
          26 0e4aa019c4690010         1642                         55          1650                       8              False
          29     1030A004F9RG          252  

In [10]:
bundle = ap.build_archive_bundle(cfg, store=store, val_stride=1, verbose=True)


grouped split by drive id -- no drive appears in two splits:
  train  280,870 drives  (12,623 carry a failure, 4.49%)
  val     49,566 drives  (2,228 carry a failure, 4.50%)
  test    82,609 drives  (3,713 carry a failure, 4.49%)
  scaler cached -- matrix already standardised in place

Window datasets (window_days=30, horizon_days=30, stride_days=1):
  train | random spd 4   |    1,123,480 windows | 280,870 drives | positive rate  1.1078% (measured, 12,623 failing drives)


  val   | enumerate stride 1  |   53,778,473 windows |  49,566 drives | positive rate  0.1277% (68,656 positive windows)
  test  | enumerate stride 1  |   89,261,367 windows |  82,609 drives | positive rate  0.1287% (114,854 positive windows)

Raw pos_weight from the training positive rate: 89.3


---
## 4 — Training, then a verdict for every 30-day window

A 1D-CNN over the sequence: dilated temporal blocks so the receptive field spans the
full 30 days, then global pooling to one logit per window.
`(batch, 30, channels) → (batch, 1)`.

**One imbalance correction, not two.** The sampler forces 25% of the windows drawn from
failing drives to be positive ones, and the loss is plain BCE. Correcting twice — a
balanced sampler *and* a large `pos_weight` — is the mistake documented in
`TOSHIBA_PIPELINE.md`.

**Early stopping on validation PR-AUC.** At a base rate near 0.1%, accuracy is
meaningless: predicting "healthy" every time scores over 99.8%.

### Choosing the architecture by experiment, not by assumption

The CNN's shape was previously a guess: five temporal blocks, 64 channels, dropout 0.2.
`scripts/run_cnn_search.py` replaces the guess with a search. It is handed a **list of
options** -- `ap.CNN_SEARCH_SPACE` -- and runs one short training run per configuration,
ranking them on the **same** validation windows:

| option | values offered |
|---|---|
| `dilations` | 2 to 6 blocks: `1-2`, `1-2-4`, `1-2-4-8`, `1-2-4-8-16`, `1-2-4-8-16-32`, `1-1-2-2-4-4` |
| `hidden_dim` | 32, 64, 96, 128 |
| `dropout` | 0.1, 0.2, 0.3 |
| `lr` | 3e-4, 1e-3, 3e-3 |
| `kernel_size` | 3, 5 |
| `weight_decay` | 1e-5, 1e-4 |
| `train_positive_ratio` | 0.15, 0.25, 0.40 |

Each `dilations` entry is one temporal block, so the search **adds and removes layers**
as well as resizing them. The full grid is 2,592 combinations; the run below is a random
sample of it, which explores each individual knob far better than an exhaustive sweep of
a few of them would.

The cell loads the cached result if the search has already been run; pass
`--trials`/`--epochs` to `scripts/run_cnn_search.py` to redo it.

In [11]:
SEARCH_CSV = "results/w2022x_cnn_search.csv"
if os.path.exists(SEARCH_CSV):
    search = pd.read_csv(SEARCH_CSV)
    print(f"loaded {len(search)} cached trials from {SEARCH_CSV}")
else:
    search = ap.cnn_search(cfg, store, bundle.splits, mode="random", n_trials=30,
                           epochs=3, val_stride=30, seed=SEED, out_csv=SEARCH_CSV)

show = ["layers", "dilations", "kernel_size", "hidden_dim", "dropout", "lr",
        "weight_decay", "train_positive_ratio", "parameters", "val_pr_auc",
        "val_roc_auc"]
print()
print("top 10 configurations by validation PR-AUC:")
print()
print(search[show].head(10).to_string(index=False))
print()
print("how depth alone scores (median over the trials at each layer count):")
print(search.groupby("layers")["val_pr_auc"].agg(["count", "median", "max"])
      .round(4).to_string())

loaded 30 cached trials from results/w2022x_cnn_search.csv

top 10 configurations by validation PR-AUC:

 layers     dilations  kernel_size  hidden_dim  dropout    lr  weight_decay  train_positive_ratio  parameters  val_pr_auc  val_roc_auc
      5    1-2-4-8-16            3         128      0.1 0.001       0.00001                  0.25      519297    0.145695     0.877631
      3         1-2-4            3          96      0.3 0.003       0.00001                  0.40      185569    0.144153     0.880100
      4       1-2-4-8            5          96      0.1 0.001       0.00010                  0.40      379297    0.143606     0.878669
      6 1-2-4-8-16-32            5         128      0.2 0.001       0.00001                  0.40      990593    0.143038     0.884129
      6   1-1-2-2-4-4            3          64      0.1 0.003       0.00001                  0.25      161729    0.142635     0.876482
      6 1-2-4-8-16-32            3         128      0.2 0.003       0.00001          

In [12]:
best = search.iloc[0]
BEST_DILATIONS = tuple(int(d) for d in str(best.dilations).split("-"))
print(f"winner: {int(best.layers)} layers, dilations {best.dilations}, "
      f"kernel {int(best.kernel_size)}, hidden {int(best.hidden_dim)}, "
      f"dropout {best.dropout}, lr {best.lr:g}, wd {best.weight_decay:g}, "
      f"positive_ratio {best.train_positive_ratio}")
print(f"  {int(best.parameters):,} parameters, search val PR-AUC {best.val_pr_auc:.4f}")

prev = search[(search.layers == 5) & (search.hidden_dim == 64)
              & (search.kernel_size == 3)]
if len(prev):
    print(f"  the previous hand-picked shape (5 layers, hidden 64, kernel 3) scored "
          f"{prev.val_pr_auc.max():.4f} -> "
          f"{100*(best.val_pr_auc/prev.val_pr_auc.max()-1):+.1f}%")

# The sampler's positive ratio was searched too, so the training set is rebuilt if the
# winner disagrees with the ratio the bundle was constructed with.
if float(best.train_positive_ratio) != cfg.train_positive_ratio:
    bundle.datasets["train"] = ap.RandomWindows(
        store, bundle.splits["train"], samples_per_drive=cfg.samples_per_drive,
        positive_ratio=float(best.train_positive_ratio), batch_size=cfg.batch_size,
        seed=SEED, split_name="train")
    print(f"  rebuilt the training sampler at positive_ratio "
          f"{float(best.train_positive_ratio)} (was {cfg.train_positive_ratio})")

winner: 5 layers, dilations 1-2-4-8-16, kernel 3, hidden 128, dropout 0.1, lr 0.001, wd 1e-05, positive_ratio 0.25
  519,297 parameters, search val PR-AUC 0.1457
  the previous hand-picked shape (5 layers, hidden 64, kernel 3) scored 0.1382 -> +5.5%


In [13]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
# Architecture and optimiser settings come from the search above, not from a guess.
tcfg = bw.TrainConfig(
    arch="cnn", hidden_dim=int(best.hidden_dim), dropout=float(best.dropout),
    dilations=BEST_DILATIONS, kernel_size=int(best.kernel_size),
    batch_size=cfg.batch_size, lr=float(best.lr),
    weight_decay=float(best.weight_decay), max_epochs=6, patience=2,
    loss="bce", sampler="none",
    # num_workers=0 on purpose: a batch is ~23 MB and worker IPC costs several times
    # what the gather itself does -- 8 workers measured 7x slower than none.
    num_workers=0, seed=SEED,
)

loaders = ap.build_loaders(bundle, tcfg, device)
model = bw.build_model(bundle.num_features, cfg.window_days, tcfg).to(device)
with torch.no_grad():
    probe = torch.zeros(2, cfg.window_days, bundle.num_features, device=device)
    print(f"input (2, {cfg.window_days}, {bundle.num_features}) -> "
          f"{tuple(model(probe).shape)}  (one logit per window)")
print(f"{bw.count_parameters(model):,} trainable parameters on {device}")

input (2, 30, 46) -> (2, 1)  (one logit per window)
519,297 trainable parameters on cuda


In [14]:
model, history, best_val = ap.train_archive_model(model, loaders, bundle, tcfg, device)
print(f"\nBest validation PR-AUC {best_val:.4f} at epoch "
      f"{1 + int(np.nanargmax(history['val_pr_auc']))}")
bw.plot_training_history(history, save_path=f"{FIG_DIR}/2022x_step3_training_history.png")

Loss: BCEWithLogitsLoss(pos_weight=1.0)  [requested]


TRAINING -- CNN, 519,297 parameters, up to 6 epochs, early stop after 2 without a val PR-AUC gain
  275 training batches/epoch, 13,130 validation batches/epoch
epoch | train loss |  val loss | val PR-AUC | val ROC-AUC | val F1@.5 |       lr |     sec
------+------------+-----------+------------+-------------+-----------+----------+--------


      epoch 1 training  20.0%  loss 0.09130


      epoch 1 training  40.0%  loss 0.07031


      epoch 1 training  60.0%  loss 0.06259


      epoch 1 training  80.0%  loss 0.05941


    1 |    0.05700 |   0.01733 |     0.1292 |      0.8798 |    0.2164 | 1.00e-03 |   518.1  <- best


      epoch 2 training  20.0%  loss 0.04775


      epoch 2 training  40.0%  loss 0.04781


      epoch 2 training  60.0%  loss 0.04775


      epoch 2 training  80.0%  loss 0.04721


    2 |    0.04712 |   0.01502 |     0.1400 |      0.8851 |    0.2332 | 1.00e-03 |   744.1  <- best


      epoch 3 training  20.0%  loss 0.04700


      epoch 3 training  40.0%  loss 0.04711


      epoch 3 training  60.0%  loss 0.04686


      epoch 3 training  80.0%  loss 0.04681


    3 |    0.04690 |   0.01560 |     0.1419 |      0.8778 |    0.2219 | 1.00e-03 |   738.2  <- best


      epoch 4 training  20.0%  loss 0.04666


      epoch 4 training  40.0%  loss 0.04654


      epoch 4 training  60.0%  loss 0.04657


      epoch 4 training  80.0%  loss 0.04641


    4 |    0.04651 |   0.01757 |     0.1478 |      0.8842 |    0.2179 | 1.00e-03 |  1326.5  <- best


      epoch 5 training  20.0%  loss 0.04669


      epoch 5 training  40.0%  loss 0.04717


      epoch 5 training  60.0%  loss 0.04684


      epoch 5 training  80.0%  loss 0.04676


    5 |    0.04636 |   0.01776 |     0.1457 |      0.8824 |    0.2290 | 1.00e-03 |  1215.7


      epoch 6 training  20.0%  loss 0.04585


      epoch 6 training  40.0%  loss 0.04524


      epoch 6 training  60.0%  loss 0.04569


      epoch 6 training  80.0%  loss 0.04550


    6 |    0.04573 |   0.01442 |     0.1541 |      0.8848 |    0.2362 | 1.00e-03 |  1316.1  <- best



Best validation PR-AUC 0.1541 at epoch 6


  saved results/figures/2022x_step3_training_history.png


C:\Users\tempuser2\Desktop\XGBOOST_2.0\XGBOOST\scripts\backblaze_window_pipeline.py:2310: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


<Figure size 1650x462 with 3 Axes>

In [15]:
results = ap.evaluate_archive_final(model, loaders, bundle, tcfg, device,
                                    pk_ks=(10, 25, 50, 100, 250, 1000))
results["history"] = history
ap.plot_evaluation(results, ks=(10, 25, 50, 100, 250, 1000),
                   save_path=f"{FIG_DIR}/2022x_step3_evaluation.png")

FINAL VALIDATION PASS -- every window at stride 1


    val: 1/13,130 batches (  0.0%)  14,265 windows/s


    val: 657/13,130 batches (  5.0%)  49,118 windows/s


    val: 1,313/13,130 batches ( 10.0%)  47,510 windows/s


    val: 1,969/13,130 batches ( 15.0%)  46,768 windows/s


    val: 2,625/13,130 batches ( 20.0%)  46,895 windows/s


    val: 3,281/13,130 batches ( 25.0%)  47,033 windows/s


    val: 3,937/13,130 batches ( 30.0%)  46,965 windows/s


    val: 4,593/13,130 batches ( 35.0%)  47,270 windows/s


    val: 5,249/13,130 batches ( 40.0%)  47,419 windows/s


    val: 5,905/13,130 batches ( 45.0%)  47,419 windows/s


    val: 6,561/13,130 batches ( 50.0%)  47,193 windows/s


    val: 7,217/13,130 batches ( 55.0%)  47,230 windows/s


    val: 7,873/13,130 batches ( 60.0%)  47,259 windows/s


    val: 8,529/13,130 batches ( 65.0%)  47,253 windows/s


    val: 9,185/13,130 batches ( 70.0%)  47,240 windows/s


    val: 9,841/13,130 batches ( 75.0%)  47,301 windows/s


    val: 10,497/13,130 batches ( 79.9%)  47,349 windows/s


    val: 11,153/13,130 batches ( 84.9%)  47,760 windows/s


    val: 11,809/13,130 batches ( 89.9%)  48,774 windows/s


    val: 12,465/13,130 batches ( 94.9%)  49,144 windows/s


    val: 13,121/13,130 batches ( 99.9%)  49,068 windows/s


  best F1 threshold .... 0.608922  (val F1 0.2511) over 53,778,473 windows

TEST PASS -- every window at stride 1


    test: 1/21,793 batches (  0.0%)  10,610 windows/s


    test: 1,090/21,793 batches (  5.0%)  42,592 windows/s


    test: 2,179/21,793 batches ( 10.0%)  42,887 windows/s


    test: 3,268/21,793 batches ( 15.0%)  42,798 windows/s


    test: 4,357/21,793 batches ( 20.0%)  42,928 windows/s


    test: 5,446/21,793 batches ( 25.0%)  44,776 windows/s


    test: 6,535/21,793 batches ( 30.0%)  48,607 windows/s


    test: 7,624/21,793 batches ( 35.0%)  51,727 windows/s


    test: 8,713/21,793 batches ( 40.0%)  54,298 windows/s


    test: 9,802/21,793 batches ( 45.0%)  56,531 windows/s


    test: 10,891/21,793 batches ( 50.0%)  58,526 windows/s


    test: 11,980/21,793 batches ( 55.0%)  60,181 windows/s


    test: 13,069/21,793 batches ( 60.0%)  61,701 windows/s


    test: 14,158/21,793 batches ( 65.0%)  63,024 windows/s


    test: 15,247/21,793 batches ( 70.0%)  62,932 windows/s


    test: 16,336/21,793 batches ( 75.0%)  60,517 windows/s


    test: 17,425/21,793 batches ( 80.0%)  58,614 windows/s


    test: 18,514/21,793 batches ( 85.0%)  56,905 windows/s


    test: 19,603/21,793 batches ( 90.0%)  55,573 windows/s


    test: 20,692/21,793 batches ( 94.9%)  54,310 windows/s


    test: 21,781/21,793 batches ( 99.9%)  54,126 windows/s


TEST @ threshold 0.5
  windows evaluated .... 89,261,367  (114,854 positive, base rate 0.129%)
  PR-AUC ............... 0.1478   (a random ranker scores 0.0013)
  ROC-AUC .............. 0.8856
  threshold ............ 0.5000
  precision ............ 0.1738
  recall ............... 0.3225
  F1 ................... 0.2259
  confusion ............ TP 37037  FP 176033  FN 77817  TN 88970480
TEST @ tuned threshold 0.608922
  windows evaluated .... 89,261,367  (114,854 positive, base rate 0.129%)
  PR-AUC ............... 0.1478   (a random ranker scores 0.0013)
  ROC-AUC .............. 0.8856
  threshold ............ 0.6089
  precision ............ 0.2318
  recall ............... 0.2588
  F1 ................... 0.2446
  confusion ............ TP 29725  FP 98504  FN 85129  TN 89048009
TEST CONFUSION MATRIX @ 0.608922   (threshold 0.6089)
                       predicted 0    predicted 1
    actual 0       89,048,009         98,504   <- 98,504 false alarms
    actual 1           85,129         

TEST PRECISION@K -- ranking windows
  89,261,367 items ranked, 114,854 positive (base rate 0.129%)
       K |   hits |  precision |    best |  recall@K |     lift
  -------+--------+------------+---------+-----------+---------
      10 |      8 |     0.8000 |  1.0000 |    0.0001 |   621.7x
      25 |     20 |     0.8000 |  1.0000 |    0.0002 |   621.7x
      50 |     36 |     0.7200 |  1.0000 |    0.0003 |   559.6x
     100 |     73 |     0.7300 |  1.0000 |    0.0006 |   567.3x
     250 |    180 |     0.7200 |  1.0000 |    0.0016 |   559.6x
    1000 |    702 |     0.7020 |  1.0000 |    0.0061 |   545.6x


TEST PRECISION@K -- ranking DRIVES (each drive scored by its worst window)
  82,609 items ranked, 3,713 positive (base rate 4.495%)
       K |   hits |  precision |    best |  recall@K |     lift
  -------+--------+------------+---------+-----------+---------
      10 |      8 |     0.8000 |  1.0000 |    0.0022 |    17.8x
      25 |     20 |     0.8000 |  1.0000 |    0.0054 |    17.8x
      50 |     38 |     0.7600 |  1.0000 |    0.0102 |    16.9x
     100 |     83 |     0.8300 |  1.0000 |    0.0224 |    18.5x
     250 |    221 |     0.8840 |  1.0000 |    0.0595 |    19.7x
    1000 |    791 |     0.7910 |  1.0000 |    0.2130 |    17.6x


SUMMARY
  val   @tuned           PR-AUC 0.1541 | ROC-AUC 0.8848 | P 0.2347 | R 0.2699 | F1 0.2511 | thr 0.6089 | TP 18530 FP 60406 FN 50126 TN 53649411
  test  @0.50            PR-AUC 0.1478 | ROC-AUC 0.8856 | P 0.1738 | R 0.3225 | F1 0.2259 | thr 0.5000 | TP 37037 FP 176033 FN 77817 TN 88970480
  test  @tuned           PR-AUC 0.1478 | ROC-AUC 0.8856 | P 0.2318 | R 0.2588 | F1 0.2446 | thr 0.6089 | TP 29725 FP 98504 FN 85129 TN 89048009


  curves drawn from 4,000,000 of 89,261,367 test windows (uniform subsample; all printed metrics use the full pass)


  saved results/figures/2022x_step3_evaluation.png


C:\Users\tempuser2\Desktop\XGBOOST_2.0\XGBOOST\scripts\backblaze_window_pipeline.py:2385: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


<Figure size 2090x462 with 4 Axes>

### The test phase, one verdict per window

The threshold is the only quantity carried out of validation, and it was fixed before the
test split was touched. Below, every tiling window of every test drive gets a single
call — **did this drive fail within 30 days of this window ending, yes or no** — scored
against the truth.

This is the tiling window set, so each test drive-day is judged exactly once.

In [16]:
thr = results["best_threshold"]
pred = ap.predict_windows(model, tiles["test"], device, threshold=thr, tag="test")

tp = int(((pred.predicted_fail == 1) & (pred.actual_fail == 1)).sum())
fp = int(((pred.predicted_fail == 1) & (pred.actual_fail == 0)).sum())
fn = int(((pred.predicted_fail == 0) & (pred.actual_fail == 1)).sum())
tn = int(((pred.predicted_fail == 0) & (pred.actual_fail == 0)).sum())

print(f"\nPER-WINDOW VERDICTS -- {len(pred):,} tiling windows over "
      f"{pred.drive_index.nunique():,} test drives, threshold {thr:.4f}\n")
print(f"  predicted FAIL ....... {int(pred.predicted_fail.sum()):,}")
print(f"  actually  FAIL ....... {int(pred.actual_fail.sum()):,}")
print(f"  correct .............. {tp + tn:,} of {len(pred):,} "
      f"({100*(tp+tn)/len(pred):.4f}% accuracy)")
print(f"  TP {tp:,}   FP {fp:,}   FN {fn:,}   TN {tn:,}")
print(f"  precision {tp/max(tp+fp,1):.4f}   recall {tp/max(tp+fn,1):.4f}")
print(f"\n  note: 'accuracy' here is dominated by the {100*(1-pred.actual_fail.mean()):.3f}% "
      f"of windows that are negative -- always saying 'no fail' would score "
      f"{100*(1-pred.actual_fail.mean()):.4f}%.")

    test: 1/755 batches (  0.1%, 1s)


    test: 76/755 batches ( 10.1%, 7s)


    test: 151/755 batches ( 20.0%, 14s)


    test: 226/755 batches ( 29.9%, 19s)


    test: 301/755 batches ( 39.9%, 24s)


    test: 376/755 batches ( 49.8%, 30s)


    test: 451/755 batches ( 59.7%, 37s)


    test: 526/755 batches ( 69.7%, 44s)


    test: 601/755 batches ( 79.6%, 51s)


    test: 676/755 batches ( 89.5%, 57s)


    test: 751/755 batches ( 99.5%, 62s)



PER-WINDOW VERDICTS -- 3,088,649 tiling windows over 82,609 test drives, threshold 0.6089

  predicted FAIL ....... 6,196
  actually  FAIL ....... 7,424
  correct .............. 3,080,373 of 3,088,649 (99.7321% accuracy)
  TP 2,672   FP 3,524   FN 4,752   TN 3,077,701
  precision 0.4312   recall 0.3599

  note: 'accuracy' here is dominated by the 99.760% of windows that are negative -- always saying 'no fail' would score 99.7596%.


In [17]:
# Every window of a few drives that really failed: the verdict as the run produced it.
failed_drives = (pred[pred.actual_fail == 1].drive_index.drop_duplicates().head(3))
for gidx in failed_drives:
    sub = pred[pred.drive_index == gidx].sort_values("window_start")
    print(f"\ndrive {sub.serial_number.iloc[0]}  --  {len(sub)} windows "
          f"covering {sub.window_start.min():%Y-%m-%d} .. {sub.window_end.max():%Y-%m-%d}")
    show = sub[["window_in_drive", "window_start", "window_end", "risk",
                "predicted_fail", "actual_fail"]].copy()
    show["verdict"] = np.where(show.predicted_fail == 1, "FAIL", "no fail")
    show["truth"] = np.where(show.actual_fail == 1, "FAIL", "no fail")
    print(show.tail(12)[["window_in_drive", "window_start", "window_end",
                         "risk", "verdict", "truth"]].to_string(index=False))


drive 1040A00VF97G  --  36 windows covering 2022-01-01 .. 2024-12-01
 window_in_drive window_start window_end     risk verdict   truth
              24   2023-12-22 2024-01-20 0.003406 no fail no fail
              25   2024-01-21 2024-02-19 0.003372 no fail no fail
              26   2024-02-20 2024-03-20 0.002816 no fail no fail
              27   2024-03-21 2024-04-19 0.004011 no fail no fail
              28   2024-04-20 2024-05-19 0.002298 no fail no fail
              29   2024-05-20 2024-06-18 0.004165 no fail no fail
              30   2024-06-19 2024-07-18 0.003385 no fail no fail
              31   2024-07-19 2024-08-17 0.003948 no fail no fail
              32   2024-08-18 2024-09-16 0.002615 no fail no fail
              33   2024-09-17 2024-10-16 0.002721 no fail no fail
              34   2024-10-17 2024-11-15 0.005162 no fail    FAIL
              35   2024-11-02 2024-12-01 0.003164 no fail    FAIL

drive 1040A0C9F97G  --  8 windows covering 2022-01-01 .. 2022-08-01
 wi

In [18]:
fig, picks = ap.plot_drive_histograms(
    pred, n_drives=24, threshold=thr, seed=SEED,
    save_path=f"{FIG_DIR}/2022x_step4_drive_histograms.png")
print(f"\n{len(picks)} drives plotted, one histogram each: the distribution of that "
      f"drive's 30-day windows over predicted risk.")

  saved results/figures/2022x_step4_drive_histograms.png

24 drives plotted, one histogram each: the distribution of that drive's 30-day windows over predicted risk.


C:\Users\tempuser2\Desktop\XGBOOST_2.0\XGBOOST\scripts\archive_window_pipeline.py:1965: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


---
## 5 — XGBoost on the same windows, head to head with the CNN

A tree model cannot take a `(30, 46)` tensor, so each **tiling** window is reduced to a
row of summary features — per channel, its last day, its 30-day mean, and its
first-to-last change: 46 channels × 3 = 138 columns. The table is built for all three
splits, from the same drive-grouped split the CNN used, so no drive crosses a boundary.

**How XGBoost is trained.**

* `tree_method="hist"` on the GPU when there is one: the 10.5M-row training table fits in
  a few minutes rather than hours.
* **Early stopping on validation PR-AUC** (`eval_metric="aucpr"`), the same criterion the
  CNN stops on, with 100 rounds of patience.
* **No imbalance weighting.** PR-AUC and Precision@K only depend on the ranking, and the
  decision threshold is tuned on validation — so `scale_pos_weight` would add a second
  correction on top of the threshold for no gain.
* **The threshold comes from validation only**, by best F1, then is frozen for test —
  exactly how the CNN's threshold was chosen.

**How the comparison is kept fair.** Both models are scored on the **same 3.1M tiling
test windows** of the same 82k test drives: the CNN's scores are the per-window verdicts
from §4, and the cell asserts that the windows and labels line up row for row. XGBoost
is not scored on the 89M stride-1 windows the CNN was also tested on in §4 — flattening
those would take ~50 GB — so the table below, not the §4 summary, is the like-for-like
read.

In [19]:
t0 = time.time()
Xtr, ytr, feat_cols = ap.window_table_features(
    store, tiles["train"], aggregates=("last", "mean", "delta"))
Xva, yva, _ = ap.window_table_features(
    store, tiles["val"], aggregates=("last", "mean", "delta"))
Xte, yte, _ = ap.window_table_features(
    store, tiles["test"], aggregates=("last", "mean", "delta"))
print(f"\nbuilt in {time.time()-t0:.0f}s   train {Xtr.shape}   val {Xva.shape}   "
      f"test {Xte.shape}")
print(f"{len(feat_cols)} features = {len(store.feature_names)} channels x "
      f"3 aggregates (last, mean, delta)")

    features 500,000/10,519,256 (  4.8%, 13s)


    features 2,500,000/10,519,256 ( 23.8%, 68s)


    features 4,500,000/10,519,256 ( 42.8%, 177s)


    features 6,500,000/10,519,256 ( 61.8%, 244s)


    features 8,500,000/10,519,256 ( 80.8%, 310s)


    features 10,500,000/10,519,256 ( 99.8%, 399s)


  feature table 10,519,256 x 138 float32 (5.8 GB), 25,254 positive (0.2401%)


    features 500,000/1,860,545 ( 26.9%, 21s)


  feature table 1,860,545 x 138 float32 (1.0 GB), 4,449 positive (0.2391%)


    features 500,000/3,088,649 ( 16.2%, 25s)


    features 2,500,000/3,088,649 ( 80.9%, 44s)


  feature table 3,088,649 x 138 float32 (1.7 GB), 7,424 positive (0.2404%)

built in 545s   train (10519256, 138)   val (1860545, 138)   test (3088649, 138)
138 features = 46 channels x 3 aggregates (last, mean, delta)


In [20]:
# The CNN is finished with the GPU; hand its cached blocks back before XGBoost copies
# the training table over.
torch.cuda.empty_cache()
xgb_device = "cuda" if torch.cuda.is_available() else "cpu"

xgb_clf = xgb.XGBClassifier(
    n_estimators=3000, learning_rate=0.05, max_depth=8, min_child_weight=5,
    subsample=0.8, colsample_bytree=0.8, max_bin=256,
    tree_method="hist", device=xgb_device,
    objective="binary:logistic", eval_metric="aucpr",
    early_stopping_rounds=100, random_state=SEED,
)
t0 = time.time()
xgb_clf.fit(Xtr, ytr, eval_set=[(Xva, yva)], verbose=100)
print(f"\nbest iteration {xgb_clf.best_iteration} of {xgb_clf.n_estimators} "
      f"(val PR-AUC {xgb_clf.best_score:.4f})  --  {time.time()-t0:.0f}s on {xgb_device}")

[0]	validation_0-aucpr:0.23294


[100]	validation_0-aucpr:0.37452


[200]	validation_0-aucpr:0.38596


[300]	validation_0-aucpr:0.39020


[400]	validation_0-aucpr:0.39178


[500]	validation_0-aucpr:0.39295


[569]	validation_0-aucpr:0.39289



best iteration 469 of 3000 (val PR-AUC 0.3932)  --  53s on cuda


In [21]:
# Predict on the CPU: the tables are NumPy arrays in host memory, and scoring them with a
# GPU booster makes XGBoost warn and fall back to a copy anyway.
xgb_clf.set_params(device="cpu")
p_va = xgb_clf.predict_proba(Xva)[:, 1]
p_te = xgb_clf.predict_proba(Xte)[:, 1]

# Threshold from validation only, then frozen -- the same rule the CNN followed.
thr_xgb, f1_va = ap.sweep_threshold(yva, p_va)
xgb_val = bw.compute_metrics(yva, p_va, thr_xgb)
xgb_half = bw.compute_metrics(yte, p_te, 0.5)
xgb_test = bw.compute_metrics(yte, p_te, thr_xgb)
print(f"threshold tuned on validation: {thr_xgb:.4f}  (val F1 {f1_va:.4f})\n")
for name, m in (("val   @tuned", xgb_val), ("test  @0.50", xgb_half),
                ("test  @tuned", xgb_test)):
    print("  " + bw.format_metrics(name, m))
print()

KS = (10, 25, 50, 100, 250, 1000)
g_te, _ = tiles["test"].locate(np.arange(tiles["test"].n_windows, dtype=np.int64))
xgb_pk_win = bw.print_precision_at_k(
    yte, p_te, "XGBOOST TEST PRECISION@K -- ranking windows", KS)
drive_codes, p_drv_xgb, y_drv = ap.drive_level_scores(p_te, yte, g_te)
xgb_pk_drv = bw.print_precision_at_k(
    y_drv, p_drv_xgb, "XGBOOST TEST PRECISION@K -- ranking DRIVES (each by its worst window)",
    KS)

imp_xgb = pd.DataFrame({"feature": feat_cols, "gain": xgb_clf.feature_importances_})
imp_xgb = imp_xgb.sort_values("gain", ascending=False).head(15).reset_index(drop=True)
imp_xgb.index += 1
print("\nwhat XGBoost relies on most (share of total gain):")
print(imp_xgb.to_string())

threshold tuned on validation: 0.2172  (val F1 0.4408)

  val   @tuned           PR-AUC 0.3932 | ROC-AUC 0.9205 | P 0.4900 | R 0.4005 | F1 0.4408 | thr 0.2172 | TP 1782 FP 1855 FN 2667 TN 1854241
  test  @0.50            PR-AUC 0.3875 | ROC-AUC 0.9186 | P 0.6992 | R 0.2411 | F1 0.3586 | thr 0.5000 | TP 1790 FP 770 FN 5634 TN 3080455
  test  @tuned           PR-AUC 0.3875 | ROC-AUC 0.9186 | P 0.4950 | R 0.3968 | F1 0.4405 | thr 0.2172 | TP 2946 FP 3006 FN 4478 TN 3078219



XGBOOST TEST PRECISION@K -- ranking windows
  3,088,649 items ranked, 7,424 positive (base rate 0.240%)
       K |   hits |  precision |    best |  recall@K |     lift
  -------+--------+------------+---------+-----------+---------
      10 |     10 |     1.0000 |  1.0000 |    0.0013 |   416.0x
      25 |     24 |     0.9600 |  1.0000 |    0.0032 |   399.4x
      50 |     49 |     0.9800 |  1.0000 |    0.0066 |   407.7x
     100 |     98 |     0.9800 |  1.0000 |    0.0132 |   407.7x
     250 |    232 |     0.9280 |  1.0000 |    0.0312 |   386.1x
    1000 |    857 |     0.8570 |  1.0000 |    0.1154 |   356.5x


XGBOOST TEST PRECISION@K -- ranking DRIVES (each by its worst window)
  82,609 items ranked, 3,713 positive (base rate 4.495%)
       K |   hits |  precision |    best |  recall@K |     lift
  -------+--------+------------+---------+-----------+---------
      10 |     10 |     1.0000 |  1.0000 |    0.0027 |    22.2x
      25 |     24 |     0.9600 |  1.0000 |    0.0065 |    21.4x
      50 |     49 |     0.9800 |  1.0000 |    0.0132 |    21.8x
     100 |     99 |     0.9900 |  1.0000 |    0.0267 |    22.0x
     250 |    241 |     0.9640 |  1.0000 |    0.0649 |    21.4x
    1000 |    921 |     0.9210 |  1.0000 |    0.2480 |    20.5x

what XGBoost relies on most (share of total gain):
                   feature      gain
1   smart_197_raw_d7__last  0.365464
2     smart_5_raw_d1__last  0.151739
3       smart_5_raw__delta  0.044702
4   smart_197_raw_d7__mean  0.039978
5     smart_5_raw_d7__mean  0.037540
6     smart_4_raw_d1__mean  0.019799
7      smart_197_raw__mean  0.017686
8        smar

In [22]:
# `pred` holds the CNN's score for exactly these tiling windows (§4). Same windows, same
# order, same labels -- asserted, not assumed.
assert len(pred) == len(yte)
assert np.array_equal(pred.drive_index.to_numpy(), g_te)
assert np.array_equal(pred.actual_fail.to_numpy(), yte.astype(int))
p_cnn = pred.risk.to_numpy()

cnn_test = bw.compute_metrics(yte, p_cnn, thr)
_, p_drv_cnn, _ = ap.drive_level_scores(p_cnn, yte, g_te)
cnn_pk_win = bw.precision_at_k_table(yte, p_cnn, KS)
cnn_pk_drv = bw.precision_at_k_table(y_drv, p_drv_cnn, KS)

comparison = pd.DataFrame([
    {"model": name, "PR-AUC": m["pr_auc"], "ROC-AUC": m["roc_auc"],
     "threshold": m["threshold"], "precision": m["precision"], "recall": m["recall"],
     "F1": m["f1"], "TP": m["tp"], "FP": m["fp"], "FN": m["fn"]}
    for name, m in (("CNN", cnn_test), ("XGBoost", xgb_test))
]).set_index("model")

print(f"CNN vs XGBOOST -- the same {len(yte):,} tiling test windows over "
      f"{len(y_drv):,} test drives")
print(f"(base rate {100*yte.mean():.3f}% of windows; each model at its own "
      f"validation-tuned threshold)\n")
print(comparison.to_string(float_format=lambda v: f"{v:.4f}"))

pk_compare = pd.DataFrame({
    "K": list(KS),
    "CNN drive hits": cnn_pk_drv.hits.to_numpy(),
    "XGBoost drive hits": xgb_pk_drv.hits.to_numpy(),
    "CNN drive precision": cnn_pk_drv.precision.to_numpy(),
    "XGBoost drive precision": xgb_pk_drv.precision.to_numpy(),
    "CNN window precision": cnn_pk_win.precision.to_numpy(),
    "XGBoost window precision": xgb_pk_win.precision.to_numpy(),
})
print(f"\nPrecision@K -- of the K drives (or windows) each model flags first, "
      f"how many really failed:\n")
print(pk_compare.to_string(index=False, float_format=lambda v: f"{v:.4f}"))

CNN vs XGBOOST -- the same 3,088,649 tiling test windows over 82,609 test drives
(base rate 0.240% of windows; each model at its own validation-tuned threshold)

         PR-AUC  ROC-AUC  threshold  precision  recall     F1    TP    FP    FN
model                                                                          
CNN      0.3232   0.9053     0.6089     0.4312  0.3599 0.3924  2672  3524  4752
XGBoost  0.3875   0.9186     0.2172     0.4950  0.3968 0.4405  2946  3006  4478

Precision@K -- of the K drives (or windows) each model flags first, how many really failed:

   K  CNN drive hits  XGBoost drive hits  CNN drive precision  XGBoost drive precision  CNN window precision  XGBoost window precision
  10               9                  10               0.9000                   1.0000                0.9000                    1.0000
  25              22                  24               0.8800                   0.9600                0.8800                    0.9600
  50              4

In [23]:
from sklearn.metrics import precision_recall_curve

fig, axes = plt.subplots(1, 2, figsize=(13, 4.2))
ax = axes[0]
for name, p, col in (("CNN", p_cnn, "#3b6ea5"), ("XGBoost", p_te, "#c1442e")):
    prec, rec, _ = precision_recall_curve(yte, p)
    ax.plot(rec, prec, color=col, lw=1.2,
            label=f"{name}  (PR-AUC {comparison.loc[name, 'PR-AUC']:.4f})")
ax.axhline(yte.mean(), color="0.5", ls="--", lw=0.8,
           label=f"random ({100*yte.mean():.2f}%)")
ax.set_xlabel("recall"); ax.set_ylabel("precision")
ax.set_title("Precision-recall, test tiling windows")
ax.legend(fontsize=8)

ax = axes[1]
ax.plot(KS, pk_compare["CNN drive precision"], "o-", color="#3b6ea5", label="CNN")
ax.plot(KS, pk_compare["XGBoost drive precision"], "o-", color="#c1442e",
        label="XGBoost")
ax.set_xscale("log"); ax.set_xticks(KS); ax.set_xticklabels([str(k) for k in KS])
ax.set_ylim(0, 1.02)
ax.set_xlabel("K drives flagged"); ax.set_ylabel("precision@K")
ax.set_title("Drives: of the top K flagged, the share that failed")
ax.legend(fontsize=8)
for a in axes:
    a.grid(alpha=0.25, linewidth=0.6); a.set_axisbelow(True)
fig.tight_layout()
fig.savefig(f"{FIG_DIR}/2022x_step5_cnn_vs_xgboost.png", dpi=130, bbox_inches="tight")
print(f"  saved {FIG_DIR}/2022x_step5_cnn_vs_xgboost.png")
plt.show()

del Xva, p_va; gc.collect()

  saved results/figures/2022x_step5_cnn_vs_xgboost.png


C:\Users\tempuser2\AppData\Local\Temp\8\ipykernel_145348\78711109.py:29: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


189

---
## 6 — A decision tree grown until it is perfect on its training set

Fitted on the same flattened **tiling** table as XGBoost, so the tree sees every
training drive-day exactly once. This is the slowest section of the notebook — a single
scikit-learn tree is fitted on one CPU core, five times over — which is why it runs last.

The depth ladder below is the point of the section. A tree with no depth limit and
`min_samples_leaf=1` will reach **1.000 on the data it was fitted to** by carving a leaf
per sample; that is what "as big as needed for 100%" produces, and the test column beside
it is what it costs. Both are printed at every depth so the gap is visible rather than
asserted.

Two things to read carefully:

* **Accuracy is the wrong lens at this base rate.** Predicting "never fails" scores about
  99.9%. Balanced accuracy — the mean of per-class recall — is reported alongside, and it
  is the column that moves.
* **100% may be unreachable and that is informative.** If two windows have identical
  summary features but different labels, no tree can separate them. The cell reports how
  many such conflicts exist, which is the ceiling on training accuracy.

In [24]:
# The exact ceiling: rows identical in every feature but disagreeing on the label can
# never be separated by any tree, however deep it is allowed to grow.
ceiling = ap.count_label_conflicts(Xtr, ytr)

  39 groups of identical feature rows carry both labels (13,981 rows, 134s)
  46 rows can never be classified correctly by any tree
  => ceiling on training accuracy: 0.99999563


In [25]:
tree, ladder = ap.grow_tree_to_full_accuracy(
    Xtr, ytr, Xte, yte, depths=[4, 8, 16, 32], seed=SEED)
print()
print(ladder.to_string(index=False))

  depth    4 -> actual   4,        16 leaves | train acc 0.997777 (bal 0.5552) | test acc 0.997766 (bal 0.5538) | 343s


  depth    8 -> actual   8,       232 leaves | train acc 0.997925 (bal 0.6065) | test acc 0.997805 (bal 0.5936) | 678s


  depth   16 -> actual  16,     2,997 leaves | train acc 0.998491 (bal 0.6917) | test acc 0.997671 (bal 0.6057) | 1322s


  depth   32 -> actual  32,     8,186 leaves | train acc 0.999008 (bal 0.7941) | test acc 0.997295 (bal 0.6218) | 2753s


  depth none -> actual 123,    28,881 leaves | train acc 0.999996 (bal 0.9991) | test acc 0.996040 (bal 0.6320) | 5891s



max_depth  actual_depth  leaves  train_accuracy  test_accuracy  train_balanced_acc  test_balanced_acc  fit_seconds
        4             4      16        0.997777       0.997766            0.555234           0.553834   342.813823
        8             8     232        0.997925       0.997805            0.606543           0.593628   678.483665
       16            16    2997        0.998491       0.997671            0.691718           0.605655  1321.534391
       32            32    8186        0.999008       0.997295            0.794130           0.621793  2752.896112
     none           123   28881        0.999996       0.996040            0.999129           0.632049  5891.139536


In [26]:
final = ladder.iloc[-1]
print(f"FINAL TREE -- depth {final.actual_depth}, {final.leaves:,} leaves, "
      f"no depth limit, min_samples_leaf=1")
print(f"  training accuracy .... {final.train_accuracy:.8f}")
print(f"  test accuracy ........ {final.test_accuracy:.8f}")
print(f"  training balanced .... {final.train_balanced_acc:.4f}")
print(f"  test balanced ........ {final.test_balanced_acc:.4f}")
print()
if final.train_accuracy >= 1.0:
    print("  reached 100.000000% on the training set: one leaf per distinguishable row.")
else:
    missed = int(round((1 - final.train_accuracy) * len(ytr)))
    print(f"  it lands {missed:,} rows short of 100%, and those are the "
          f"{ceiling['unreachable_rows']:,} rows that are")
    print(f"  IDENTICAL across all {Xtr.shape[1]} features to a row carrying the "
          f"opposite label. No tree can")
    print(f"  split what it cannot tell apart, so "
          f"{100*ceiling['max_train_accuracy']:.6f}% is the ceiling -- the tree is")
    print(f"  already at it, not falling short of it.")
print(f"\n  the gap between the two ACCURACY columns is the tree memorising "
      f"{final.leaves:,} leaves of training rows;")
print(f"  the gap between the two BALANCED columns "
      f"({final.train_balanced_acc:.3f} vs {final.test_balanced_acc:.3f}) is what that "
      f"memorisation is")
print(f"  actually worth on a disk it has never seen.\n")

imp = pd.DataFrame({"feature": feat_cols, "importance": tree.feature_importances_})
imp = imp[imp.importance > 0].sort_values("importance", ascending=False).head(15)
imp.index = range(1, len(imp) + 1)
print("what the tree splits on most:")
print(imp.to_string())

FINAL TREE -- depth 123, 28,881 leaves, no depth limit, min_samples_leaf=1
  training accuracy .... 0.99999563
  test accuracy ........ 0.99604034
  training balanced .... 0.9991
  test balanced ........ 0.6320

  it lands 46 rows short of 100%, and those are the 46 rows that are
  IDENTICAL across all 138 features to a row carrying the opposite label. No tree can
  split what it cannot tell apart, so 99.999563% is the ceiling -- the tree is
  already at it, not falling short of it.

  the gap between the two ACCURACY columns is the tree memorising 28,881 leaves of training rows;
  the gap between the two BALANCED columns (0.999 vs 0.632) is what that memorisation is
  actually worth on a disk it has never seen.

what the tree splits on most:
                    feature  importance
1    smart_197_raw_d7__last    0.071871
2      smart_5_raw_d1__last    0.037265
3       smart_194_raw__mean    0.025644
4     smart_9_raw_d7__delta    0.022574
5   smart_193_raw_d7__delta    0.019868
6     s

In [27]:
fig, axes = plt.subplots(1, 2, figsize=(13, 4.2))
ax = axes[0]
x = np.arange(len(ladder))
ax.plot(x, ladder.train_accuracy, "o-", color="#3b6ea5", label="train accuracy")
ax.plot(x, ladder.test_accuracy, "o-", color="#c1442e", label="test accuracy")
ax.set_xticks(x); ax.set_xticklabels(ladder.max_depth.astype(str))
ax.set_xlabel("max_depth"); ax.set_ylabel("accuracy")
ax.set_title("Accuracy is the wrong lens here\n(both columns pinned near 1.0 by the base rate)")
ax.legend(fontsize=8)

ax = axes[1]
ax.plot(x, ladder.train_balanced_acc, "o-", color="#3b6ea5", label="train balanced acc")
ax.plot(x, ladder.test_balanced_acc, "o-", color="#c1442e", label="test balanced acc")
ax.set_xticks(x); ax.set_xticklabels(ladder.max_depth.astype(str))
ax.set_xlabel("max_depth"); ax.set_ylabel("balanced accuracy")
ax.set_title("Balanced accuracy: the gap that opens is the overfitting")
ax.legend(fontsize=8)
for a in axes:
    a.grid(alpha=0.25, linewidth=0.6); a.set_axisbelow(True)
fig.tight_layout()
fig.savefig(f"{FIG_DIR}/2022x_step6_tree_depth.png", dpi=130, bbox_inches="tight")
print(f"  saved {FIG_DIR}/2022x_step6_tree_depth.png")
plt.show()

  saved results/figures/2022x_step6_tree_depth.png


C:\Users\tempuser2\AppData\Local\Temp\8\ipykernel_145348\607113841.py:23: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [28]:
os.makedirs("models", exist_ok=True); os.makedirs("results", exist_ok=True)
ap.save_artifacts(model, bundle, tcfg, results, prefix="models/backblaze_2022x_w30")
xgb_clf.save_model("models/backblaze_2022x_w30_xgb.json")
with open("models/backblaze_2022x_w30_xgb_config.json", "w", encoding="utf-8") as fh:
    json.dump({"feature_columns": feat_cols, "aggregates": ["last", "mean", "delta"],
               "window_days": cfg.window_days, "horizon_days": cfg.horizon_days,
               "threshold": thr_xgb, "best_iteration": int(xgb_clf.best_iteration),
               "params": {k: v for k, v in xgb_clf.get_params().items()
                          if isinstance(v, (int, float, str, bool)) or v is None}},
              fh, indent=2)
print("Saved models/backblaze_2022x_w30_xgb.json and its config")

comparison.to_csv("results/w2022x_cnn_vs_xgboost.csv")
pk_compare.to_csv("results/w2022x_cnn_vs_xgboost_precision_at_k.csv", index=False)
imp_xgb.to_csv("results/w2022x_xgboost_importance.csv")

pd.DataFrame(history).to_csv("results/w2022x_training_history.csv", index=False)
coverage.to_csv("results/w2022x_window_coverage.csv")
cov.to_csv("results/w2022x_per_drive_windows.csv", index=False)
ladder.to_csv("results/w2022x_tree_depth_ladder.csv", index=False)
results["top_drives"].to_csv("results/w2022x_top_drives.csv", index=False)
checks["channel_scores"].to_csv("results/w2022x_channel_label_auc.csv", index=False)
pred.head(2_000_000).to_csv("results/w2022x_window_verdicts.csv", index=False)
pd.concat([results["precision_at_k_windows"].assign(ranked="windows"),
           results["precision_at_k_drives"].assign(ranked="drives")],
          ignore_index=True).to_csv("results/w2022x_precision_at_k.csv", index=False)

with open("results/w2022x_summary.json", "w", encoding="utf-8") as fh:
    json.dump({
        "since_year": SINCE_YEAR,
        "shards": [os.path.basename(p)[:-8] for p in shards],
        "features": list(features),
        "blocklist": report.blocklist,
        "n_drives": int(store.n_drives),
        "n_drive_days": int(store.n_rows),
        "tiling_windows_total": int(total_tiles),
        "tiling_batches_total": int(total_batches),
        "leakage_checks_passed": bool(checks["x_independent_of_label"]),
        "max_channel_auc_vs_label": float(checks["max_channel_auc"]),
        "best_threshold": results["best_threshold"],
        "test_at_tuned": results["test_at_tuned"],
        "xgboost": {"best_iteration": int(xgb_clf.best_iteration),
                    "threshold": thr_xgb,
                    "val_at_tuned": xgb_val,
                    "test_at_tuned": xgb_test,
                    "test_at_half": xgb_half},
        "cnn_on_tiling_test": cnn_test,
        "tree_ceiling": {k: float(v) for k, v in ceiling.items()},
        "tree_final": {k: (float(v) if isinstance(v, (int, float, np.floating)) else str(v))
                       for k, v in ladder.iloc[-1].to_dict().items()},
    }, fh, indent=2)
print("wrote results/w2022x_*.csv and results/w2022x_summary.json")

Saved models/backblaze_2022x_w30_cnn.pth and models/backblaze_2022x_w30_preprocessing_config.json


Saved models/backblaze_2022x_w30_xgb.json and its config


wrote results/w2022x_*.csv and results/w2022x_summary.json


---
## Summary — CNN vs XGBoost vs decision tree

All three models on the **same 3.1M tiling test windows** of the same 82k test drives.
The CNN and XGBoost use the threshold each tuned on validation; the tree never saw
validation, so it is cut at 0.5.

The winner is named on two criteria, in order of importance:

1. **drive precision @100** — of the 100 drives a model flags first, the share that
   really failed. This is the decision an operator makes.
2. **PR-AUC** — the quality of the whole ranking.

When the top two sit within 5% of each other on a criterion, it is reported as a tie
rather than a win.

In [29]:
p_tree = tree.predict_proba(Xte)[:, 1]
tree_test = bw.compute_metrics(yte, p_tree, 0.5)
_, p_drv_tree, _ = ap.drive_level_scores(p_tree, yte, g_te)
tree_pk_drv = bw.precision_at_k_table(y_drv, p_drv_tree, KS)

entries = [
    ("CNN", cnn_test, cnn_pk_drv),
    ("XGBoost", xgb_test, xgb_pk_drv),
    (f"Decision tree (depth {int(final.actual_depth)})", tree_test, tree_pk_drv),
]
rows = []
for name, m, pk in entries:
    pk = pk.set_index("k")
    rows.append({
        "model": name,
        "drive precision @100": pk.loc[100, "precision"],
        "drives hit @100": int(pk.loc[100, "hits"]),
        "drives hit @1000": int(pk.loc[1000, "hits"]),
        "PR-AUC": m["pr_auc"],
        "ROC-AUC": m["roc_auc"],
        "precision": m["precision"],
        "recall": m["recall"],
        "F1": m["f1"],
        "threshold": m["threshold"],
    })
summary = pd.DataFrame(rows).set_index("model")


def verdict(col):
    ranked = summary[col].sort_values(ascending=False)
    (first, a), (second, b) = list(ranked.items())[:2]
    if a <= 0 or (a - b) / a < 0.05:
        return f"{first} and {second} are tied ({a:.4f} vs {b:.4f})"
    return f"{first} ({a:.4f}, next best {second} at {b:.4f})"


rule = "=" * 88
print(rule)
print(f"SUMMARY -- {len(yte):,} test windows over {len(y_drv):,} test drives "
      f"({int(y_drv.sum()):,} failed; window base rate {100*yte.mean():.3f}%)")
print(rule)
print(summary.to_string(float_format=lambda v: f"{v:.4f}"))
print()
print(f"  best at flagging drives (precision @100) ... {verdict('drive precision @100')}")
print(f"  best ranking overall (PR-AUC) ............. {verdict('PR-AUC')}")
print()
print(f"  for reference, not comparable to the table: the CNN on every stride-1 test "
      f"window (§4) scored")
print(f"  PR-AUC {results['test_at_tuned']['pr_auc']:.4f} over "
      f"{results['n_test_windows']:,} overlapping windows.")

summary.to_csv("results/w2022x_model_summary.csv")
print("\nwrote results/w2022x_model_summary.csv")

SUMMARY -- 3,088,649 test windows over 82,609 test drives (3,713 failed; window base rate 0.240%)
                           drive precision @100  drives hit @100  drives hit @1000  PR-AUC  ROC-AUC  precision  recall     F1  threshold
model                                                                                                                                   
CNN                                      0.9200               92               856  0.3232   0.9053     0.4312  0.3599 0.3924     0.6089
XGBoost                                  0.9900               99               921  0.3875   0.9186     0.4950  0.3968 0.4405     0.2172
Decision tree (depth 123)                0.2300               23               350  0.0619   0.6318     0.2252  0.2664 0.2441     0.5000

  best at flagging drives (precision @100) ... XGBoost (0.9900, next best CNN at 0.9200)
  best ranking overall (PR-AUC) ............. XGBoost (0.3875, next best CNN at 0.3232)

  for reference, not comparable to the

---
## What these numbers do and do not say

**The label never reached the model.** §2 is an executed experiment, not a promise: the
labels were replaced with noise and the window tensors came back bit-identical. Together
with the audit's blocklist, that is the strongest statement available short of a proof —
the model's ranking comes from SMART telemetry and nothing else.

**The ranking transfers; the base rate is already real.** This is the fleet at its own
prevalence, not a matched cohort, so the precision figures need no discounting before
they are read. Read the **lift** column in Precision@K rather than raw precision, and
prefer the drive-level table — an operator pulls *K disks*, not K windows.

**Tiling and stride-1 answer different questions.** The tiling set judges every
drive-day exactly once, which is what makes the per-window verdict table and the tree
honest. The stride-1 set scores every window that exists, which is the right basis for a
ranking metric. Neither is a sample of the other.

**CNN vs XGBoost is a comparison of inputs as much as of models.** The CNN reads all
30 days of every channel in order; XGBoost reads three numbers per channel. If XGBoost
matches the CNN, most of the signal is in the level and trend of a few SMART counters
and the day-by-day shape adds little. If the CNN wins, the order of events inside the
window carries information the summaries throw away. Read the drive-level Precision@K
rows first — that is the decision an operator makes — and PR-AUC second.

**The 100% tree is a memorisation result, not a model.** It reaches perfect training
accuracy by growing a leaf per training row, and the test column next to it is the honest
read. At a 0.1% base rate even the test accuracy is near-perfect while the model is
nearly useless — which is exactly why every other number in this notebook is PR-AUC,
Precision@K or balanced accuracy rather than accuracy.

**2022 onward is not the whole archive.** Drive models, capacities and the SMART
attributes actually populated all shift over time; a model fitted on the last few years
is answering a narrower and more current question than one fitted on all thirteen.